### DISCUSSION TITLE
Vercel Rewrite: SPA Fallback except for API endpoints

### DISCUSSION THREAD

**Discussion by ricokahler** [vercel/vercel#5448] [Help]
I asked [this question](https://stackoverflow.com/questions/64920230/vercel-rewrite-spa-fallback-except-for-api-endpoint) on stackoverflow but copying here for visibility. If someone answers here, I'll copy the answer over there

---

I have a single page application + API I'm deploying to Vercel.

I'm currently trying to add some configuration in my vercel.json that:

1. redirects calls not at the root (e.g. `/login) to `index.html` so that I can utilize the HTML History API (i.e. react-router's browser router)
2. expect not for the `/api` endpoint where I have a few dynamic paths (e.g. `/files/[fileId]/[checksum].ts`)

How do I create a rewrite in Vercel that accomplishes this?

I've followed the advice from here: https://vercel.com/docs/configuration#routes/advanced/spa-fallback

```json
{
  "rewrites": [{ "source": "/(.*)", "destination": "/index.html" }]
}
```

However this does redirects API endpoints with dynamic paths to the index.html file too where I just want this to be the API.

Any ideas?

**Comment by cesargdm** [comment]
Still facing the same problem

**Comment by VanyaBerzoi** [comment]
> I asked [this question](https://stackoverflow.com/questions/64920230/vercel-rewrite-spa-fallback-except-for-api-endpoint) on stackoverflow but copying here for visibility. If someone answers here, I'll copy the answer over there
> 
> ---
> 
> I have a single page application + API I'm deploying to Vercel.
> 
> I'm currently trying to add some configuration in my vercel.json that:
> 
> 1. redirects calls not at the root (e.g. `/login) to `index.html` so that I can utilize the HTML History API (i.e. react-router's browser router)
> 2. expect not for the `/api` endpoint where I have a few dynamic paths (e.g. `/files/[fileId]/[checksum].ts`)
> 
> How do I create a rewrite in Vercel that accomplishes this?
> 
> I've followed the advice from here: https://vercel.com/docs/configuration#routes/advanced/spa-fallback
> 
> ```json
> {
>   "rewrites": [{ "source": "/(.*)", "destination": "/index.html" }]
> }
> ```
> 
> However this does redirects API endpoints with dynamic paths to the index.html file too where I just want this to be the API.
> 
> Any ideas?
>

**Comment by VanyaBerzoi** [comment]
G

**Reply by VanyaBerzoi** [reply]
-

**Reply by VanyaBerzoi** [reply]
> - 

- [ ] - @![Uploading c{2ae5aa9} (94b66ea8-6247-4107-9b9f-2e02874af42c id=0x7f0a03c2 tag=BaseCommentFragment) …]()

**Reply by devinhalladay** [reply]
@VanyaBerzoi can you re-link to the comment above? The link seems to be broken now.

**Comment by devinhalladay** [comment]
Same problem here! Curiously, my API routes work just fine in production, but not locally when using `vercel dev`. API routes are totally inaccessible locally, and seem to be taken over by `BrowserRouter` from `react-router-dom@6.3.0`.

```tsx
// /src/App.tsx
<BrowserRouter>
  <Routes>
    <Route path="/" element={<Home />} />
    {hasLibrary && (
      <>
        <Route path="/library" element={<Library />} />
        <Route path="/library/:id" element={<Article />} />
      </>
    )}
    <Route element={<p>not found</p>} />
  </Routes>
</BrowserRouter>;
```

```js
// /api/reader.js
export default (request, response) => {
  const { name } = request.query;
  response.status(200).send(`Hello ${name}!`);
};
```

Like @ricokahler, I followed the docs [here](https://vercel.com/docs/project-configuration#legacy/routes/upgrading) to no avail.

```json
// Vercel.json
{
  "rewrites": [
    {
      "source": "/((?!api\/.*).*)",
      "destination": "/index.html"
    }
  ]
}
```

I also tried an alternate from the docs, which didn't work:

```json
// Vercel.json
{
  "routes": [
    { "handle": "filesystem" },
    {
      "src": "/((?!api/.*).*)",
      "dest": "/index.html"
    }
  ]
}
```

Sadly I've found no docs from react router 6, or even on substack, pointing to a solution.

**Reply by devinhalladay** [reply]
Related discussion here: https://github.com/vercel/vercel/issues/3602

**Comment by nhowell** [comment]
Not sure if it'll help in the case of rewriting everything _except_ `/api/*`, but I found that the `destination` had to be set to `/` instead of `/index.html`.

For example, in my use case:
```json
"rewrites": [{ "source": "/(.*)", "destination": "/" }]

**Comment by devotox** [comment]
@ricokahler i think this should work using the new [negative lookahead syntax](https://vercel.com/docs/errors/error-list#invalid-route-source-pattern)

```
  "rewrites": [
    {
      "source": "/((?!api\/.*).*)",
      "destination": "/index.html"
    }
  ],


  "headers": [
    {
      "source": "/((?!api).*)",
      "headers": [
        {
          "key": "Cache-Control",
          "value": "public, max-age=86400, s-maxage=86400, stale-while-revalidate"
        }
      ]
    }
  ]
```
